File: src/readxl/zip.cpp
#pragma once
#include "zip.h"

#include "cpp4r/function.hpp"
#include "cpp4r/raws.hpp"

std::string zip_buffer(const std::string &zip_path,
                       const std::string &file_path) {
  cpp4r::function zip_buffer = cpp4r::package("readODS")["zip_buffer"];

  cpp4r::raws xml(zip_buffer(zip_path, file_path));
  std::string buffer(RAW(xml), RAW(xml) + xml.size());
  buffer.push_back('\0');

  return buffer;
}

bool zip_has_file(const std::string &zip_path, const std::string &file_path) {
  cpp4r::function zip_has_file = cpp4r::package("readODS")["zip_has_file"];
  return zip_has_file(zip_path, file_path);
}

File: R/zip.R
# Taken from readxl

# Called only from C++ code, but currently needs to be implemented in R.
zip_buffer <- function(zip_path, file_path) {
    files <- zip::zip_list(zip_path)
    indx <- match(file_path, files$filename)
    if (is.na(indx)) {
        stop("Couldn't find '", file_path, "' in '", zip_path, "'", call. = FALSE)
    }
    size <- files$uncompressed_size[indx]
    con <- unz(zip_path, file_path, open = "rb")
    on.exit(close(con), add = TRUE)
    readBin(con, raw(), n = size)
}

zip_has_file <- function(zip_path, file_path) {
    file_path %in% zip::zip_list(zip_path)$filename
}
